runtime/src/read_optimized_dashmap.rs

```diff
@@ -0,0 +1,199 @@
+#![allow(dead_code)]
+
+#[cfg(feature = "shuttle-test")]
+use shuttle::sync::Arc;
+#[cfg(not(feature = "shuttle-test"))]
+use std::sync::Arc;
+use {
+    dashmap::{
+        mapref::{entry::Entry, multiple::RefMulti},
+        DashMap,
+    },
+    std::{hash::BuildHasher, ops::Deref},
+};
+
+type DashmapIteratorItem<'a, K, V, S> = RefMulti<'a, K, ROValue<V>, S>;
+
+// Wrapper around dashmap that stores (K, Arc<V>) to minimize shard contention.
+#[derive(Debug)]
+pub struct ReadOptimizedDashMap<K, V, S>
+where
+    K: Clone + Eq + std::hash::Hash,
+    S: Clone + BuildHasher,
+{
+    inner: DashMap<K, ROValue<V>, S>,
+}
+
+impl<K, V, S> ReadOptimizedDashMap<K, V, S>
+where
+    K: Clone + Eq + std::hash::Hash,
+    S: Clone + BuildHasher,
+{
+    pub fn new(inner: DashMap<K, ROValue<V>, S>) -> Self {
+        Self { inner }
+    }
+
+    /// Alternative to entry(k).or_insert_with(default) that returns an Arc<V> instead of returning a
+    /// guard that holds the underlying shard's write lock.
+    ///
+    /// # Safety
+    ///
+    /// - Care must be taken when mutating the returned value. If modifications are applied while
```

Comment:
sorry for nit on documentation and placement of unsafe. Seems like the `unsafe` should be on accessing the `ROValue` since that's where the safety comes in.
This would add an clear reason for the `ROValue` struct imo, as it makes us document `unsafe` access on those.
I think this fn is actually unconditionally safe.

Does that make sense to you, or just in my head?

Reply:
removed unsafe